File: python/heapSort.py
class MaxHeap:
    def __init__(self, array):
        self.heap = array
    
    def buildHeap(self):
        startIdx = (len(self.heap)-2) // 2
        while startIdx >= 0:
            self.siftDown(startIdx, len(self.heap)-1)
            startIdx -= 1
            
        print("After Initial Build -> ", self.heap)
            
    def biggest(self, x, y):
        if self.heap[x] >= self.heap[y]:
            return x
        return y
    
    def remove(self, window):
        endIdx = len(self.heap) - window - 1
        self.heap[0], self.heap[endIdx] = self.heap[endIdx], self.heap[0]
        self.siftDown(0, endIdx-1)
        
    def siftDown(self, startIdx, endIdx):
        while startIdx < endIdx:
            firstChild, secondChild = (startIdx * 2) + 1, (startIdx * 2) + 2
            if firstChild > endIdx:
                break
            biggest = 0
            if secondChild <= endIdx:
                biggest = self.biggest(firstChild, secondChild)
            else:
                biggest = firstChild
                
            if self.heap[biggest] > self.heap[startIdx]:
                self.heap[biggest], self.heap[startIdx] = self.heap[startIdx], self.heap[biggest]
                startIdx = biggest
            else:
                break
            
     
# Time -> 0(N(log(N))) Space -> 0(1) 
def heapSort(array):
    heap = MaxHeap(array)
    heap.buildHeap()
    idx = 0
    
    while idx < len(array) - 1:
        heap.remove(idx)
        idx += 1
        
    return array

print(heapSort([8, 5, 2, 9, 5, 6, 3]))

File: go/heapSort/heapSort.go
package main

import "fmt"

type MaxHeap []int 

func NewMaxHeap(array []int) *MaxHeap {
	heap := MaxHeap(array)
	ptr := &heap
	ptr.BuildHeap(array)
	fmt.Println("After Initial Build -> ", heap)
	return ptr
}

func (heap *MaxHeap) BuildHeap(array []int) {
	startIdx := (len(array) - 2) / 2
	for startIdx >= 0 {
		heap.SiftDown(startIdx, len(array)-1)
		startIdx--
	}
}

func (heap *MaxHeap) Bigest(x, y int) int {
	if (*heap)[x] >= (*heap)[y] {
		return x
	}
	return y
}

func (heap *MaxHeap) Remove(window int) {
	endIdx := len(*heap) - window - 1
	(*heap)[0], (*heap)[endIdx] = (*heap)[endIdx], (*heap)[0]
	heap.SiftDown(0, endIdx-1)
}

func (heap *MaxHeap) SiftDown(startIdx, endIdx int) {
	for startIdx < endIdx {
		firstChild, secondChild := (startIdx * 2) + 1, (startIdx * 2) + 2
		if firstChild > endIdx {
			break
		}
		var smallest int
		if secondChild <= endIdx {
			smallest = heap.Bigest(firstChild, secondChild)
		}else {
			smallest = firstChild
		}

		if (*heap)[smallest] > (*heap)[startIdx] {
			(*heap)[smallest], (*heap)[startIdx] = (*heap)[startIdx], (*heap)[smallest]
			startIdx = smallest
		}else {
			break
		}
	}
}


// Time -> 0(N(log(N))) Space -> 0(1)
func HeapSort(array []int) []int {
	heap := NewMaxHeap(array)

	for idx:=0; idx < len(array)-1; idx++ {
		heap.Remove(idx)
	}
	return array
}


func main() {
	fmt.Println(HeapSort([]int{8, 5, 2, 9, 5, 6, 3}))
}